Enhance Window: add autoDestroy and autoCenter properties

derrell opened the pull request:
If the autoDestroy property is true, automatically remove a window from its root container, remove all listeners and contents, and dispose the window, when it is closed.

If the autoCenter property is true, automatically center the window when it is opened.

There is no change to prior behavior, by default, with this PR. Both autoDestroy and autoCenter default to false.

derrell committed:
Enhance Window: add autoDestroy and autoCenter properties

cajus commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -632,17 +679,34 @@ qx.Class.define("qx.ui.window.Window",
+        qx.core.Init.getApplication().getRoot().remove(this);// remove self
+        this.dispose();
+      }
     },
```
For some reason we did not do it this way and used the *onDisappear* event to handle that. I don't recall the reason, because it's so much time has passed now. Will check when I'm back @ work...

level420 replied:
I think the removal from the focus handler is missing here:

```javascript
qx.ui.core.FocusHandler.getInstance().removeRoot(this);
```

level420 replied:
What about packing the destruction code into one overridable method. This way deriving classes could add their own code or even interrupt auto destruction in some cases.

oetiker commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -632,17 +679,34 @@ qx.Class.define("qx.ui.window.Window",
+
+      // If autoCenter is true, then center the window
+      if (this.getAutoCenter()) {
+        this.center();
```
at least in my code I usually center in response to the appear event ... and recenter on resize ... as discussed on gitter

level420 replied:
The same with our code. I've created a mixin which adds a ```centered``` property and adds an event listener to both ```resize``` and ```appear``` event.
Additionally I've added an applier to the ```centered``` property which centers the window when set to true while the window is open.

level420 replied:
The same with our code. I've created a mixin which adds a ```centered``` property and adds an event listener to both ```resize``` and ```appear``` event.
Additionally I've added an applier to the centered property which centers the window when set to true while the window is open.

johnspackman commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -632,17 +679,34 @@ qx.Class.define("qx.ui.window.Window",
+        this.removeAll();                                    // remove contents
+        qx.event.Registration.removeAllListeners(this);      // remove listeners
+        qx.core.Init.getApplication().getRoot().remove(this);// remove self
+        this.dispose();
```
Can this be reduced to just `this.dispose()`?  `this.removeAll()` is done by the `Widget.destruct` already, and `qx.event.Registration.removeAllListeners` is handled by `Object.destruct`.  

I would expect that `getRoot().remove()` does need to be called, but in the current implementation it looks like it's *never* called, even if the Window is manually destroyed -- which would sound like a bug you'd expect to have cropped up by now, so maybe it's already handled elsewhere somehow?

However, if something needs to be done during destruction/disposal, the code could be moved into the destructor or overridden `dispose()` method so it's always implemented

level420 replied:
Yes it should be reduced to a ```this.dispose()``` and by adding additional code into the ```destruct``` method. The additional code could be
```javascript
qx.ui.core.FocusHandler.getInstance().removeRoot(this);
qx.core.Init.getApplication().getRoot().remove(this);
```

level420 replied:
I've digged through the framework code without finding any call to ```qx.core.Init.getApplication().getRoot().remove``` there, so I think windows are added but never removed from application root.

derrell replied:
The fact that I couldn't find anyplace that removed the window from the root was what started me on this path.

Thanks for all of the comments, guys. I'll update this PR in the coming days.

derrell committed:
Further enhance Window
- Moved much of the autoDestroy functionality into the destructor, and
  allow the normal dispose mechanism to handle as much as possible
- Provided capability to automatically center on window appear and/or
  resize. Ensured that parent's resize handler that centered the window got
  removed upon window destruction.

Derrell

johnspackman commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -993,6 +1018,112 @@ qx.Class.define("qx.ui.window.Window",
+        });
+
+      return values;
+    },
```
this transform is perfectly valid code, but it seems very complicated for what the end goal is trying to achieve, ie a simple pair of booleans.  How about changing the property to `check: [ "appear", "resize", "both", "never" ]`?  It's a common approach to this problem and a lot easier to maintain

derrell replied:
I try to avoid that type of thing. It has no future capability to enhance it without greatly complicating the list of options. Consider if we wanted to have center upon resize of the window as another option. Then we'd have [ "appear", "resizeParent", "resizeWindow", "appear+resizeParent", "appear+resizeWindow", "resizeWindow+resizeParent", "appear+resizeWindow+resizeParent" ] ... and you'd have to list them in the correct order. Whereas, with the current code, if we want to add a third option, it's changing only that simple switch statement.

derrell replied:
(oh, and I forgot "never" in my modified list)

johnspackman replied:
I see your point - but is it necessary to convert from strings in various forms?  The centerWhen property has a given type (array) and set of allowed values in that array, I don't think it's appropriate to coerce the values from a subset of other unknown values, especially if that means parsing strings into arrays etc.  IMHO it's the caller's requirement to use the correct type.

As a side issue, I agree with the usefulness of sets as property values, it would be great if there was a way to express this in qx property definitions but perhaps we can talk about that on gitter?

derrell replied:
BTW, users who want to extend Window can override the transform and/or apply functions to add their own keys here.

oetiker commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -372,6 +372,78 @@ qx.Class.define("qx.ui.window.Window",
+     * Examples:
+     *   These two yield identical behavior, centering upon appear and resize:
+     *     win.setCenterWhen([ "appear", "resize" ]);
+     *     win.setCenterWhen("appear resize");
```
any particular reason for whipping up a DSL here ? (albite a simple one)

derrell committed:
By popular demand, allow only array values for the 'centerWhen' property

derrell committed:
Be careful not to play with (null) parent if window has already been removed

derrell committed:
remove now-redundant check of parent

derrell committed:
Can't have multiple setLayoutParent functions. Combine the two.

cajus commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -618,12 +732,15 @@ qx.Class.define("qx.ui.window.Window",
     close : function()
     {
-      if (!this.isVisible()) {
+      if (!this.getAutoDestroy() && !this.isVisible()) {
```
Why can't I close a window where *autoDestroy* is set to *true*? Not sure what that check is good foor, can you explain?

derrell replied:
The original code just looked at isVisible because hide() and close() were essentially the same thing. If it was already not visible, through either a hide() call or a prior close() call, then there was nothing to do here. With autoDestroy true, we don't want to return. When close() is called, we definitely want it to close and be destroyed even if hide() had previously been called.

derrell replied:
Sorry, commented on this separately. See other explanation.

cajus commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -929,6 +1055,40 @@ qx.Class.define("qx.ui.window.Window",
+      if (this.__centeringAppearId !== null) {
+        this.removeListenerById(this.__centeringAppearId);
+        this.__centeringAppearId = null;
+      }
```
At least the listener on *this* could be omitted. Just calling *this.removeListener("appear", this.center, this)* is enough, as it doesn't fail when it does not exist. But - well, that's a matter of taste, of course.

level420 replied:
I think all this logic could be omitted including the dynamic removal and re-assignment of the listeners. I'll add the listeners for resize and appear anyway and ask in the listener if the corresponding property is true and center or not. I don't think it's worth all that listener tracking.

derrell replied:
That's just a preference of methodology, right? No technical issue here?

level420 replied:
Yes, just how I would have done it. Nothing technical.

cajus commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -1050,5 +1210,29 @@ qx.Class.define("qx.ui.window.Window",
+
+    // Remove the listener for appear, if there is one
+    id = this.__centeringAppearId;
+    id && this.removeListenerById(id);
```
That two lines can be omitted since listeners to this will be automatically removed.

level420 replied:
It is not necessary to track the listener id for listeners attached to ```this```

derrell replied:
Yes, I need to track the listener id so that I can remove the listener if the property value changes. (It's much faster to use the listener id than to search through the list of listeners for the one that matches a given function and context.) But yes, I can remove these two lines here because all listeners of `this` are removed in higher-layer code.

cajus reviewed (changes requested):
If I ignore the _centerWhen_ property, it looks good for me.

A proposal: if it's important to react on _appear_, _resize_, _appear/resize_, then just add two boolean properties - i.e. _autoCenterOnResize_ and _autoCenterOnAppear_. So you're getting rid of all that hassle with checking, shared Array references, etc.

level420 commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -372,6 +372,96 @@ qx.Class.define("qx.ui.window.Window",
+     *     win.setCenterWhen([]);
+     */
+    centerWhen :
+    {
```
I'm here with @cajus: It would be much simpler and understandably if we had two boolean properties ```centerOnAppear``` and ```centerOnResize```. No arrays, no initialization hassle.

oetiker replied:
I also think having just two properties would be perfectly fine and easy to understand ...

derrell replied:
I don't like it, but ok, I'll change it to two properties. It may take me a day or three before I can get to it, though.

level420 commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -929,6 +1055,40 @@ qx.Class.define("qx.ui.window.Window",
       }
     },
 
+    // overridden
```
The method is not overridden, it's new to this class, I think.

derrell replied:
apply functions are always overridden. They override the default code. Or at least that's the way it is throughout most qooxdoo code. I don't care. I'll remove that, but I think it's "the standard."

level420 replied:
Afaik applyers are not created automatically and then overriden. If you declare a property and define there an applyer attribute, then using that property via the setter will throw an exception. See http://tinyurl.com/zsd4oaq

level420 commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -929,6 +1055,40 @@ qx.Class.define("qx.ui.window.Window",
+
+      // If we are to center on resize, arrange to do so
+      if (value.indexOf("resize") != -1) {
+        parent = this.getLayoutParent();
```
It could theoretically happen that the layout parent changes, I think. We'd then have to hook into the ```setLayoutParent``` method, remove the event listener from the old parent and attach the listener to the new one. This gets simpler if we always attach a resize listener to the parent. All this can be done in ```setLayoutParent```

derrell replied:
Yes, the layout parent can change. That's why there is code in the overridden setLayoutParent method to handle that.

level420 commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -1050,5 +1210,29 @@ qx.Class.define("qx.ui.window.Window",
+
+    // If we haven't been removed from our parent, clean it up too.
+    parent = this.getLayoutParent();
+    if (parent) {
```
If the logic for attaching/detaching the resize listener to the parent is moved to ```setLayoutParent``` we could simply call here ```this.setLayoutParent(null);``` which removes the parent listener if it's there.

derrell replied:
That removes a few lines of code, while making the code more difficult to understand.

derrell replied:
Also, I still need to retrieve the parent in order to remove `this` from it.

level420 commented:
It would be nice to have a test for the "center on appear" which I think could be done by some calculations on the position within the parent.

derrell commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -372,6 +372,96 @@ qx.Class.define("qx.ui.window.Window",
+     */
+    centerWhen :
+    {
+      init : [],
```
Yes, I'll add a proper init function. Thanks. With regard to the push() scenario, if they want to try to break the code, it's their own fault. This isn't end-user usage; it's programmer usage.

derrell committed:
uniquely initialize centerWhen property; don't superfluously remove listeners

derrell committed:
Replace convenient single property for centering with two properties

oetiker commented on framework/source/class/qx/ui/window/Window.js:
```diff
@@ -929,6 +1020,43 @@ qx.Class.define("qx.ui.window.Window",
 
+    _applyCenterOnAppear : function(value, old)
+    {
+      var             parent = this.getLayoutParent();
```
this is not used anymore ...

derrell committed:
Remove obsolete variable declaration

zaucker commented:
@derrell I like your last commit message ... ;-)
